teste2e/test_format_upgrades.py

```diff
@@ -0,0 +1,116 @@
+from util import UrsadbTestContext
+from util import ursadb  # noqa
+import json
+import subprocess
+import time
+from pathlib import Path
+import tempfile
+import os
+import json
+from typing import Dict, Any
+
+
+class TemporaryStorage:
+    def __init__(self):
+        self.tmpfiles = []
+
+    def tmpfile(self, name=None) -> Path:
+        if name is None:
+            name = os.urandom(8)
+        filepath = tempfile.gettempdir() + "/" + os.urandom(8).hex()
+        self.tmpfiles.append(filepath)
+        return Path(filepath)
+
+    def write_json(self, data: Dict[str, Any], name=None) -> Path:
+        p = self.tmpfile(name=name)
+        p.write_text(json.dumps(data))
+        return p
+
+    def write_text(self, text: str, name=None) -> Path:
+        p = self.tmpfile(name=name)
+        p.write_text(text)
+        return p
+
+    def free(self):
+        for fpath in self.tmpfiles:
+            os.unlink(fpath)
+
+
+def upgrade_json(data: Dict[str, Any]) -> Dict[str, Any]:
+    """ Tries to upgrade database passed in Dict. Will create
+    a whole tree of database objects, in order to pass db checks. """
+    ursadb_root = Path(__file__).parent.parent / "build"
+    ursadb = ursadb_root / "ursadb"
+    backend = "tcp://127.0.0.1:9877"
+
+    ctx = TemporaryStorage()
+
+    db = ctx.write_json(data)
+    for dataset in data.get("datasets", []):
+        filesf = ctx.write_text("").name
+        ctx.write_json(
+            {"files": filesf, "indices": [], "taints": []}, name=dataset
+        )
+
+    if data.get("iterators") is not None:
+        for itermeta in data["iterators"]:
+            iterf = ctx.write_text("").name
+            ctx.write_json(
+                {
+                    "backing_storage": iterf,
+                    "byte_offset": 0,
+                    "file_offset": 0,
+                    "total_files": 0,
+                },
+                name=itermeta,
+            )
+
+    ursadb = subprocess.Popen([ursadb, db, backend])
+    # TODO remove this after we have some sort of quit command
+    time.sleep(1)
```

Comment:
:thinking:

Reply:
this doesn't really guarantee that the operation succeeded, maybe:
1. launch database
2. `ursacli -q -c status`
3. when the above finished - terminate the DB

ursacli is backed by zmq which will gracefully wait till the database starts to be operational (at least as far as I remember)

Reply:
:thinking: right, completely forgot that I can just ping the DB.